### PR TITLE
Smatrix tool returns multi-frequency DataArray

### PR DESCRIPTION
**Pull request by tylerflex**
Two major changes:
1. Smatrix tool returns a DataArray with dims `("port_in", "port_out", "mode_index_in", "mode_index_out", "f")`
2. Smatrix tool accepts `freqs` instead of single `freq`

### CODE REVIEW

**Comment by momchil-flex** [comment]
There's a few things I wanna work on still and Shashwat is away this week. @tylerflex you want to get this in 2.0 could you get someone else to review/test it. Maybe @lucas-flexcompute ?

**Comment by tylerflex** [comment]
Sure, sounds good. @lucas-flexcompute / @momchil-flex  So yea I'm trying to get this in for 2.0 so that we dont break backwards compatibility later with these changes. Just as a summary, the only things this PR does is:
1. Makes the matrix tool work with many frequencies (which each get added to the monitors)
2. Thee output of the Smatrix tool is a data array with fields of port name (in and out), mode_index (in and out), and frequency, instead of a nested dict, which was clumsy to work with.

I plan some further refactoring of the tool after 2.0 but these are just the main changes needed to the API side, so we dont break backwards compability once I get working on it. You can see the API changes in action at [the corresponding docs PR](https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/9b0724d029b0721b575a0763476e00ddbf83783b/docs/source/notebooks/SMatrix.ipynb)

File: tidy3d/plugins/smatrix/smatrix.py [hunk 1]
```diff
@@ -65,10 +90,10 @@ class ComponentModeler(Tidy3dBaseModel):
         description="Collection of ports describing the scattering matrix elements. "
         "For each port, one simulation will be run with a modal source.",
```
Comment (lucas-flexcompute) [inline tidy3d/plugins/smatrix/smatrix.py:91]: “For each **input mode**, one simulation will be run…”

File: tidy3d/plugins/smatrix/smatrix.py [hunk 2]
```diff
@@ -48,7 +49,31 @@ class Port(Box):
...
+    ...     port_out=ports_out,
+    ...     mode_index_in=mode_index_in,
+    ...     mode_index_out=mode_index_out,
+    ...     f=f
+    ... )
+    >>> fd = SMatrixDataArray((1 + 1j) * np.random.random((2, 2, 2, 2, 1)), coords=coords)
+    """
+
+    __slots__ = ()
+    _dims = ("port_in", "port_out", "mode_index_in", "mode_index_out", "f")
```
Comment (lucas-flexcompute) [inline tidy3d/plugins/smatrix/smatrix.py:75]: I'd change the ordering here to `("port_out", "mode_index_out", "port_in", "mode_index_in", "f")` to be more compatible with a matrix representation if we use `SMatrixDataArray.values`. The usual ordering of indices is output first, input second.

**Review by lucas-flexcompute — APPROVED** [review]
The descriptions of [`element_mappings`](https://github.com/flexcompute/tidy3d/pull/716/files#diff-8cfee9637dc09ebab5e95853d64a3466280e369f9560a2b807ba239da5290e18R112-R113) and [`run_only`](https://github.com/flexcompute/tidy3d/pull/716/files#diff-8cfee9637dc09ebab5e95853d64a3466280e369f9560a2b807ba239da5290e18R119-R120) make reference to the matrix rows and columns swapped to the conventional definition of S matrix. Given the new type class, we might as well get rid of this nomenclature altogether and say input/output only. Also, I wouldn't say "the resulting scattering matrix will not be square or complete," but maybe "the data corresponding to other inputs will be missing in the resulting matrix"

In `ComponentModeler`, is it a problem for the simulation to have monitors defined? The description says so, but the validator doesn't check that. I ask because I imagine the user wanting to look at the field distribution over the component for specific inputs after solving the S matrix.

**Comment by lucas-flexcompute** [comment]
Regarding the notebook, the coupler as it is defined is not very interesting, as nothing is actually coupled. If you change the following dimensions, you'll get a more interesting S matrix:
```python
# Waveguide width
wg_width = 0.5
# length of coupling region (um)
coup_length = 5.0
# spacing between waveguides in coupling region (um)
wg_spacing_coup = 0.07
```

**Comment by tylerflex** [comment]
> ```python
> # Waveguide width
> wg_width = 0.5
> # length of coupling region (um)
> coup_length = 5.0
> # spacing between waveguides in coupling region (um)
> wg_spacing_coup = 0.07
> ```

Thanks, I just tried with these parameters and it does look a lot better, updated the notebook!

**Comment by tylerflex** [comment]
> [`only`](https://github.com/flexcompute/tidy3d/pull/716/files#diff-8cfee9637dc09ebab5e95853d64a3466280e369f9560a2b807ba239da5290e18R119-R120) make reference to the matrix rows and columns swapped to the conventional definition of S matrix. Given the new type class, we might as well get rid of this nomenclature altogether and say input/output only. Also, I wouldn't say "the resulting scattering matrix will not be square or complete," but maybe "the data corresponding to other inputs will be missing in the resulting matrix"
> 
> In `ComponentModeler`, is it a problem for the simulation to have monitors defined? The description says so, but the validator doesn't check that. I ask because I imagine the user wanting to look at the field distribution over the component for specific inputs after solving the S matrix.

Thanks, in the version I will push, all of these comments / conventions are fixed.

**Pull request merged**